src-extras/Database/LSMTree/Extras/MergingTreeData.hs

```diff
@@ -0,0 +1,358 @@
+  where
+    go (CompletedTreeMergeData rd) =
+          tabulate "merging tree state" ["CompletedTreeMerge"]
+        . labelRunData rd
+    go (OngoingTreeMergeData mrd) =
+          tabulate "merging tree state" ["OngoingTreeMerge"]
+        . labelMergingRunData mrd
+    go (PendingLevelMergeData prds mtd) =
+          tabulate "merging tree state" ["PendingLevelMerge"]
+        . foldr ((.) . labelPreExistingRunData) id prds
+        . maybe id go mtd
+    go (PendingUnionMergeData mtds) =
+          tabulate "merging tree state" ["PendingUnionMerge"]
+        . foldr ((.) . go) id mtds
+
+    -- the longest path from the root to a run
+    depthTree = (+1) . \case  -- maximum depth of children
+        CompletedTreeMergeData _ -> 0
+        OngoingTreeMergeData _   -> 0
+        PendingLevelMergeData prds mtds ->
+          maximum (0 : fmap (const 1) prds ++ map depthTree (toList mtds))
+        PendingUnionMergeData mtds ->
+          maximum (0 : map depthTree mtds)
+
+
+labelPreExistingRunData :: SerialisedPreExistingRunData -> Property -> Property
+labelPreExistingRunData (PreExistingRunData rd)         = labelRunData rd
+labelPreExistingRunData (PreExistingMergingRunData mrd) = labelMergingRunData mrd
+
+instance ( Ord k, Arbitrary k, Arbitrary v, Arbitrary b
+         ) => Arbitrary (MergingTreeData k v b) where
+  arbitrary = genMergingTreeData arbitrary arbitrary arbitrary
+  shrink = shrinkMergingTreeData shrink shrink shrink
+
+genMergingTreeData ::
+     forall k v b. Ord k
+  => Gen k
+  -> Gen v
+  -> Gen b
+  -> Gen (MergingTreeData k v b)
+genMergingTreeData genKey genVal genBlob = QC.sized $ \s -> do
+    go =<< QC.chooseInt (1, max 1 s)
+  where
+    -- n is the number of PMergingTree constructors + pre-existing runs
+    go n
+      | n < 1
+      = error ("arbitrary T: n == " <> show n)
+
+      | n == 1
+      = QC.frequency
+          [ (1, CompletedTreeMergeData <$> genRunData genKey genVal genBlob)
+          , (1, OngoingTreeMergeData <$> genMergingRunData genKey genVal genBlob)
+          ]
+
+      | otherwise
+      = QC.frequency
+          [ (1, do
+              -- pending level merge without child
+              preExisting <- QC.vectorOf (n - 1) $  -- 1 for constructor itself
+                genPreExistingRunData genKey genVal genBlob
+              return (PendingLevelMergeData preExisting Nothing))
+          , (1, do
+              -- pending level merge with child
+              numPreExisting <- QC.chooseInt (0, min 20 (n - 2))
+              preExisting <- QC.vectorOf numPreExisting $
+                genPreExistingRunData genKey genVal genBlob
+              tree <- go (n - numPreExisting - 1)
+              return (PendingLevelMergeData preExisting (Just tree)))
+          , (2, do
+              -- pending union merge
+              ns <- QC.shuffle =<< arbitraryPartition2 n
+              PendingUnionMergeData <$> traverse go ns)
+          ]
```

Comment:
Same comment about the number of children: on't we test the most interesting cases with `2` or `3` children?

Reply:
Here it does indeed matter and I landed on distribution between 0 and 6 pre-existing run, but skewing towards smaller numbers.

Reply:
How do you feel about the `chooseIntSkewed` I went with?

Reply:
It has a noticeable, but relatively small effect. Changing the range to `(0, 3)` lead to about 25% faster tests. I decided to generate 50% shorter runs instead (see last commit), halving the runtime.